Issue with RedundantBraces and v2.5.3 : illegal start of simple expression

- **Configuration**:
```
version=2.5.3

rewrite.rules = [RedundantBraces]
```

## Steps

Given code like this:
```scala
object Issue {
  def f(): Int = ???

  try {} finally {
    val _ = f()
  }
}
```

When I run scalafmt like this:

```bash
scalafmt
```

## Problem

I got an error

```
Issue.scala:6: error: illegal start of simple expression
```

## Workaround

Setting `rewrite.redundantBraces.generalExpressions` to `false` seems to fix.
(the default value has changed from `false` to `true` in 2.5.0)

## Notes

It seems to work with version 2.4.x and fail with 2.5.x

Here is another example that fails with the same error

```
object Issue2 {
  def f(): Int = ???
  def g : Int => Unit = _ => {val _ = f()}
}
```

--- a/scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/RedundantBraces.scala
+++ b/scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/RedundantBraces.scala
@@ -232,11 +232,6 @@ class RedundantBraces(implicit ctx: RewriteCtx) extends RewriteSession {
       case d: Defn.Def =>
         def disqualifiedByUnit =
           !settings.includeUnitMethods && d.decltpe.exists(_.syntax == "Unit")
-        def innerOk(s: Stat) =
-          s match {
-            case _: Term.Function | _: Defn => false
-            case _ => true
-          }
         settings.methodBodies &&
         getSingleStatIfLineSpanOk(b).exists(innerOk) &&
         !isProcedureSyntax(d) &&
@@ -251,6 +246,12 @@ class RedundantBraces(implicit ctx: RewriteCtx) extends RewriteSession {
     }
   }
 
+  private def innerOk(s: Stat) =
+    s match {
+      case _: Term.Function | _: Defn => false
+      case _ => true
+    }
+
   private def okToRemoveBlockWithinApply(b: Term.Block): Boolean =
     getSingleStatIfLineSpanOk(b).exists {
       case f: Term.Function =>
@@ -263,7 +264,7 @@ class RedundantBraces(implicit ctx: RewriteCtx) extends RewriteSession {
   /** Some blocks look redundant but aren't */
   private def shouldRemoveSingleStatBlock(b: Term.Block): Boolean =
     getSingleStatIfLineSpanOk(b).exists { stat =>
-      !b.parent.exists {
+      innerOk(stat) && !b.parent.exists {
         case parentIf: Term.If if stat.is[Term.If] =>
           // if (a) { if (b) c } else d
           //   ↑ cannot be replaced by ↓

--- a/scalafmt-tests/src/test/resources/rewrite/RedundantBraces.stat
+++ b/scalafmt-tests/src/test/resources/rewrite/RedundantBraces.stat
@@ -845,10 +845,10 @@ object Issue {
   }
 }
 >>>
-test does not parse
-OUT: object Issue {
+object Issue {
   def f(): Int = ???
 
-  try {} finally
+  try {} finally {
     val _ = f()
   }
+}

--- a/scalafmt-tests/src/test/resources/rewrite/RedundantBraces.stat
+++ b/scalafmt-tests/src/test/resources/rewrite/RedundantBraces.stat
@@ -834,3 +834,21 @@ object a {
 object a {
   scalacOptions ~= { _ filterNot (_ startsWith "-Xlint") }
 }
+<<< #2012
+rewrite.redundantBraces.generalExpressions = true
+===
+object Issue {
+  def f(): Int = ???
+
+  try {} finally {
+    val _ = f()
+  }
+}
+>>>
+test does not parse
+OUT: object Issue {
+  def f(): Int = ???
+
+  try {} finally
+    val _ = f()
+  }

